### PR TITLE
Feature bearer token auth

### PR DESCRIPTION
**Pull request by Xeloss**
Allows bearer token authentication on HTTP Request. The plug-in will try to match the OID extracted from the token with the OID that must be stored in the users's email field.

### CODE REVIEW

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java [hunk 1]
```diff
@@ -76,69 +67,35 @@ public String getEditPropertiesJspFilePath() {
   @Nullable
   @Override
   public Collection<String> validate(@NotNull Map<String, String> properties) {
-    final Collection<String> errors = new ArrayList<String>();
+
```
Comment (jfatta) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java]: is this newline necessary?

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java [hunk 2]
```diff
@@ -76,69 +67,35 @@ public String getEditPropertiesJspFilePath() {
...
+	  final Collection<String> errors = new ArrayList<String>();
     if(StringUtil.isEmptyOrSpaces(properties.get(AADConstants.AUTH_ENDPOINT_SCHEME_PROPERTY_KEY))){
       errors.add("App OAuth 2.0 authorization endpoint should be specified.");
     }
     if(StringUtil.isEmptyOrSpaces(properties.get(AADConstants.CLIENT_ID_SCHEME_PROPERTY_KEY))){
       errors.add("Client ID should be specified.");
     }
+    
+    if(Boolean.valueOf(properties.get(AADConstants.ENABLE_TOKEN_AUTHENTICATION)) && !Boolean.valueOf(properties.get(AADConstants.ALLOW_MATCHING_USERS_BY_EMAIL))){
+      errors.add("In order to enable token authentication, matching users by email must be enable too.");
```
Comment (jfatta) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java]: *enabled

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADLoginTokenAuthenticatior.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	      LOG.warn(error);
+	      return sendUnauthorized(request, response, errorDescription);
+	    }
+	
+	    final String nonce = token.getClaim(NONCE_CLAIM);
+	    final String name = token.getClaim(NAME_CLAIM);
+	    final String oid = token.getClaim(OID_CLAIM);
+	
+	    if (nonce == null || name == null || oid == null)
+	      return sendBadRequest(response, String.format("Some of required claims were not found in parsed JWT. nonce - %s; name - %s, oid - %s", nonce, name, oid));
```
Comment (jfatta) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADLoginTokenAuthenticatior.java]: > The nonce, name and oid claims are required for Token authentication. Parsed values: ...

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADLoginTokenAuthenticatior.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	
+	    if(!nonce.equals(SessionUtil.getSessionId(request)))
+	      return sendBadRequest(response, "Marked request as unauthenticated since retrieved JWT 'nonce' claim doesn't correspond to current TeamCity session.");
```
Comment (jfatta) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADLoginTokenAuthenticatior.java]: > ... the current TeamCity session

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/BearerTokenAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+import org.jetbrains.annotations.NotNull;
+
+import jetbrains.buildServer.controllers.interceptors.auth.HttpAuthenticationResult;
+import jetbrains.buildServer.serverSide.auth.ServerPrincipal;
+import jetbrains.buildServer.web.openapi.PluginDescriptor;
+
+public class BearerTokenAuthenticator extends TokenAuthenticator {
+
+	private static final String OID_CLAIM = "oid"; //object ID
```
Comment (jfatta) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/BearerTokenAuthenticator.java]: we should keep the constants defined in one place in order to avoid duplicates

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/JWT.java [hunk 6]
```diff
@@ -32,6 +32,7 @@ private JWT(@NotNull JsonObject content) {
   @Nullable
   public static JWT parse(@NotNull String jwtString){
 	LOG.debug (String.format("Verifying JWT: %s", jwtString));
+
```
Comment (jfatta) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/JWT.java]: remove this new line

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/JWT.java [hunk 7]
```diff
@@ -59,6 +64,7 @@ public static JWT parse(@NotNull String jwtString){
       LOG.warn("Failed to parse JWT from JWS payload " + jwsPayload, e);
       return null;
     }
+
```
Comment (jfatta) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/JWT.java]: same here, revert these changes if not necessary.

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/ServerPrincipalFactory.java [hunk 8]
```diff
@@ -32,23 +31,28 @@ public ServerPrincipalFactory(@NotNull UserModel userModel) {
 
   @NotNull
   public ServerPrincipal getServerPrincipal(@NotNull String userName, @NotNull final String aadUserUID, @Nullable final String email, @NotNull Map<String, String> schemeProperties) {
-    final ServerPrincipal existingPrincipal = findExistingPrincipalByUID(aadUserUID);
-    if(existingPrincipal != null) return existingPrincipal;
+    
+	final ServerPrincipal existingPrincipal = findExistingPrincipalByUID(aadUserUID);
+    if(existingPrincipal != null)
```
Comment (jfatta) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/ServerPrincipalFactory.java]: let's use curly braces always :)

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/ServerPrincipalFactory.java [hunk 9]
```diff
@@ -32,23 +31,28 @@ public ServerPrincipalFactory(@NotNull UserModel userModel) {
...
   public ServerPrincipal getServerPrincipal(@NotNull String userName, @NotNull final String aadUserUID, @Nullable final String email, @NotNull Map<String, String> schemeProperties) {
-    final ServerPrincipal existingPrincipal = findExistingPrincipalByUID(aadUserUID);
-    if(existingPrincipal != null) return existingPrincipal;
+    
+	final ServerPrincipal existingPrincipal = findExistingPrincipalByUID(aadUserUID);
+    if(existingPrincipal != null) 
+    	return existingPrincipal;
 
     if(email != null && allowMatchUserByEmail(schemeProperties)){
+
```
Comment (jfatta) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/ServerPrincipalFactory.java]: revert unnecessary changes to ease the PR review

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java (hunk 1 above)
Comment (Xeloss) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java]: No, i took it off.

Comment (Xeloss) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java]: No, I took it off.

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java (hunk 2 above)
Comment (Xeloss) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java]: Done!

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADLoginTokenAuthenticatior.java (hunk 4 above)
Comment (Xeloss) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADLoginTokenAuthenticatior.java]: Corrected!

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADLoginTokenAuthenticatior.java (hunk 3 above)
Comment (Xeloss) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADLoginTokenAuthenticatior.java]: Changed

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/BearerTokenAuthenticator.java (hunk 5 above)
Comment (Xeloss) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/BearerTokenAuthenticator.java]: I moved all claim-related constant to a new file "ClaimsConstants"

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/JWT.java (hunk 7 above)
Comment (Xeloss) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/JWT.java]: Ok, I reverted the changes. I added them for readability purpose.

File: azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/ServerPrincipalFactory.java (hunk 8 above)
Comment (Xeloss) [inline azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/ServerPrincipalFactory.java]: Fixed
